### PR TITLE
Changes to _eval_step_func

### PR DESCRIPTION
**Pull request by kombatEldridge**
In my project, I am wanting to use some step functions while Meep runs. Currently, my entire simulation of Meep runs in a Simulation class and I would like to keep my step functions in this class as well. That being said, they must have `self` as an arg. With the current `_eval_step_func`, the arg handler is a bit too strict. There is probably a reason for this, and please disregard this PR if so.

The changes I am requesting would include the possibility that the step functions contain an arg named `self` but should still keep the intended limitations of the previous version.

### CODE REVIEW

**Comment by oskooi** [comment]
>  Currently, my entire simulation of Meep runs in a Simulation class and I would like to keep my step functions in this class as well. That being said, they must have self as an arg. With the current _eval_step_func, the arg handler is a bit too strict.

If you haven't already done so, it might be useful to first read [The Run Function is not a Loop](https://meep.readthedocs.io/en/master/The_Run_Function_Is_Not_A_Loop/) which provides a description of the step-function object including examples.

File: python/simulation.py [hunk 1]
```diff
@@ -4996,14 +5004,16 @@ def _combine(sim, todo):
 
 def _eval_step_func(sim, func, todo):
     num_args = get_num_args(func)
+    name_args = get_name_args(func)
+    self_count = int("self" in name_args)
```
Comment (stevengj) [inline python/simulation.py:5008]: Using `inspect.ismethod(func)` seems like a more reliable way to check for this.  However, this seems unreliable in another way: it assumes that any bound method is a method of the `sim` object.    What if it is a method of some other object?

**Comment by stevengj** [comment]
Note that you can already pass as many additional arguments you want to the step function by wrapping it in a [closure](https://www.pythontutorial.net/advanced-python/python-closures/).

For example, if you have a function `my_step(a,b,c,sim)`, you can pass it as a step function with `lambda sim: my_step(a,b,c,sim)`, in which the additional arguments `a,b,c` are captured by the closure.

Or, if you have a function `foo.my_step(sim)` that is a bound method of some object `foo`, you can pass it as a step function with `lambda sim: foo.my_step(sim)`.

File: python/simulation.py [hunk 2]
```diff
@@ -100,6 +100,14 @@ def get_num_args(func):
     )
 
 
+def get_name_args(func):
+    return (
+        ("sim", "todo")
+        if isinstance(func, Harminv) or isinstance(func, PadeDFT)
+        else func.__code__.co_varnames[: func.__code__.co_argcount]
+    )
```
Comment (stevengj) [inline python/simulation.py:108]: This can be eliminated if we use `inspect.ismethod` as indicated below.

**Comment by stevengj** [comment]
Overall, I'm in favor of something like this to make it easier to pass bound methods for step functions.

However, what I'd like to do is to simply fix `get_num_args` to subtract 1 if the function is a bound method, as determined by `inspect.ismethod(func)` (not from the argument names!).

File: python/simulation.py [hunk 3]
```diff
@@ -96,6 +97,8 @@ def get_num_args(func):
     return (
         2
         if isinstance(func, Harminv) or isinstance(func, PadeDFT)
+        else func.__code__.co_argcount - 1
```
Comment (stevengj) [inline python/simulation.py:100]:
```suggestion
        else func.__code__.co_argcount - 1 # first argument is "self", passed implicitly for bound methods
```

**Review by stevengj — APPROVED** [review]

File: python/simulation.py [hunk 4]
```diff
@@ -96,6 +97,8 @@ def get_num_args(func):
     return (
         2
         if isinstance(func, Harminv) or isinstance(func, PadeDFT)
+        else func.__code__.co_argcount - 1 # first argument is "self", passed implicitly for bound methods
```
Comment (stevengj) [inline python/simulation.py:100]:
```suggestion
        else func.__code__.co_argcount - 1 # first argument is "self"
```

File: python/simulation.py [hunk 5]
```diff
@@ -96,6 +97,8 @@ def get_num_args(func):
     return (
         2
         if isinstance(func, Harminv) or isinstance(func, PadeDFT)
+        else func.__code__.co_argcount - 1 # first argument is "self"
```
Comment (stevengj) [inline python/simulation.py:100]:
```suggestion
        else func.__code__.co_argcount - 1  # first argument is "self"
```

File: python/simulation.py [hunk 6]
```diff
@@ -96,6 +97,8 @@ def get_num_args(func):
     return (
         2
         if isinstance(func, Harminv) or isinstance(func, PadeDFT)
+        else func.__code__.co_argcount - 1  # first argument is "self"
+        if inspect.ismethod(func)
```
Comment (oskooi) [inline python/simulation.py:101]: For improved readability, I would recommend you replace this ternary operator which spans multiple lines with a full `if` statement. 

See: https://google.github.io/styleguide/pyguide.html#2111-definition.

**Pull request merged**